### PR TITLE
check remote attributes in local

### PR DESCRIPTION
**Pull request by YuechengLiu**
update a set for remote attributes in the local object every time __setattr__() is called

### CODE REVIEW

**Comment by zenghsh3** [comment]
add unittest cases:
1. try to get attribute after adding a new attribute in function calls
2. try to get attribute after setting an attribute

File: parl/remote/job.py [hunk 1]
```diff
@@ -397,25 +400,12 @@ def single_task(self, obj, reply_socket, job_address):
...
-                                dumps_return(True)
-                            ])
-                        else:
-                            reply_socket.send_multipart([
-                                remote_constants.NORMAL_TAG,
-                                dumps_return(False)
-                            ])
-
-                    elif tag == remote_constants.CALL_TAG:
+                    if tag == remote_constants.CALL_TAG:
```
Comment (zenghsh3) [inline parl/remote/job.py:408]: After calling function, the keys in `__dict__` may be changed. We should also return the latest keys in `__dict__` here.

Comment (YuechengLiu) [reply parl/remote/job.py:408]: fixed

File: parl/remote/remote_decorator.py [hunk 2]
```diff
@@ -190,33 +194,18 @@ def request_cpu_resource(self, global_client, max_memory):
...
                 self.internal_lock.acquire()
                 self.job_socket.send_multipart([
-                    remote_constants.SET_ATTRIBUTE,
+                    remote_constants.SET_ATTRIBUTE_TAG,
                     to_byte(attr),
                     dumps_return(value)
                 ])
                 message = self.job_socket.recv_multipart()
                 tag = message[0]
                 self.internal_lock.release()
```
Comment (zenghsh3) [inline parl/remote/remote_decorator.py]: we should release the lock after the line `self.remote_attribute_keys_set = loads_return(message[1])`

Comment (YuechengLiu) [reply parl/remote/remote_decorator.py]: modified

**Review by zenghsh3 — DISMISSED** [review]
LGTM

File: parl/remote/remote_decorator.py [hunk 3]
```diff
@@ -225,14 +214,15 @@ def set_remote_attr(self, attr, value):
             def get_remote_attr(self, attr):
                 """Call the function of the unwrapped class."""
                 #check if attr is a attribute or a function
-                is_attribute = self.check_attribute(attr)
+                is_attribute = (attr in self.remote_attribute_keys_set)
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/remote_decorator.py]: remove the bracket here.

Comment (YuechengLiu) [reply parl/remote/remote_decorator.py]: modified

**Review by TomorrowIsAnOtherDay — APPROVED** [review]

**Pull request merged**